onadata/libs/utils/logger_tools.py

```diff
@@ -158,43 +159,47 @@ def create_instance(
-        # content hash is not present, by string comparison of the full
-        # content, which is slow! Use the management command
-        # `populate_xml_hashes_for_instances` to hash existing submissions
-        existing_instance = Instance.objects.filter(
-            Q(xml_hash=xml_hash) | Q(xml_hash=Instance.DEFAULT_XML_HASH, xml=xml),
-            xform__user=xform.user,
-        ).first()
-    else:
-        existing_instance = None
-
-    if existing_instance:
-        existing_instance.check_active(force=False)
-        # ensure we have saved the extra attachments
-        new_attachments = save_attachments(existing_instance, media_files)
-        if not new_attachments:
-            raise DuplicateInstance()
+    with get_instance_lock(xml_hash) as lock_acquired:
+        if not lock_acquired:
+            raise ConflictingXMLHashInstanceError()
+
+        # Dorey's rule from 2012 (commit 890a67aa):
+        #   Ignore submission as a duplicate IFF
+        #    * a submission's XForm collects start time
+        #    * the submitted XML is an exact match with one that
+        #      has already been submitted for that user.
+        # The start-time requirement protected submissions with identical responses
+        # from being rejected as duplicates *before* KoBoCAT had the concept of
+        # submission UUIDs. Nowadays, OpenRosa requires clients to send a UUID (in
+        # `<instanceID>`) within every submission; if the incoming XML has a UUID
+        # and still exactly matches an existing submission, it's certainly a
+        # duplicate (https://docs.opendatakit.org/openrosa-metadata/#fields).
+        if xform.has_start_time or new_uuid is not None:
+            # XML matches are identified by identical content hash OR, when a
+            # content hash is not present, by string comparison of the full
+            # content, which is slow! Use the management command
+            # `populate_xml_hashes_for_instances` to hash existing submissions
+            existing_instance = Instance.objects.filter(
+                Q(xml_hash=xml_hash) | Q(xml_hash=Instance.DEFAULT_XML_HASH, xml=xml),
+                xform__user=xform.user,
+            ).first()
         else:
-            # Update Mongo via the related ParsedInstance
-            existing_instance.parsed_instance.save(asynchronous=False)
-            return existing_instance
-    else:
-        instance = save_submission(request, xform, xml, media_files, new_uuid,
-                                   status, date_created_override)
-        return instance
+            existing_instance = None
+
+        if existing_instance:
+            existing_instance.check_active(force=False)
+            # ensure we have saved the extra attachments
+            new_attachments = save_attachments(existing_instance, media_files)
+            if not new_attachments:
+                raise DuplicateInstanceError()
```

Comment:
Wouldn't `ConflictingXMLHashInstanceError` be raised already and block us from getting here?